File: core/src/main/java/vn/loitp/restapi/DateTypeDeserializer.java
package vn.loitp.restapi;

import android.util.Log;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;

/**
 * @author hoangphu
 * @since 7/21/16
 */
public class DateTypeDeserializer implements JsonDeserializer<Date> {

    private static final String TAG = DateTypeDeserializer.class.getName();
    private static final String[] DATE_FORMATS = new String[]{
            "dd-MM-yyyy hh:mm",
            "dd/MM/yyyy",
            "yyyy-MM-dd"
    };

    @Override
    public Date deserialize(JsonElement jsonElement, Type typeOF, JsonDeserializationContext context) throws JsonParseException {
        for (String format : DATE_FORMATS) {
            try {
                return new SimpleDateFormat(format, Locale.getDefault()).parse(jsonElement.getAsString());
            } catch (ParseException e) {
                Log.v(TAG, "Error when deserialize date");
            }
        }
        throw new JsonParseException("Cant parse date: \"" + jsonElement.getAsString()
                + "\". Supported formats: \n" + Arrays.toString(DATE_FORMATS));
    }
}

File: core/src/main/java/com/restapi/DateTypeDeserializer.kt
package com.restapi

import android.util.Log
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class DateTypeDeserializer : JsonDeserializer<Date> {

    companion object {
        private val TAG = DateTypeDeserializer::class.java.name
        private val DATE_FORMATS = arrayOf(
                "dd-MM-yyyy hh:mm",
                "dd/MM/yyyy",
                "yyyy-MM-dd"
        )
    }

    @Throws(JsonParseException::class)
    override fun deserialize(jsonElement: JsonElement, typeOF: Type, context: JsonDeserializationContext): Date? {
        for (format in DATE_FORMATS) {
            try {
                return SimpleDateFormat(
                        format, Locale.getDefault()
                ).parse(jsonElement.asString)
            } catch (e: ParseException) {
                Log.e(TAG, "Error when deserialize date")
            }
        }
        throw JsonParseException("""Can't parse date: "${jsonElement.asString}". Supported formats:${DATE_FORMATS.contentToString()}""".trimIndent())
    }
}
